File: frontend/src/services/cache.ts
import { openDB, DBSchema, IDBPDatabase } from 'idb';

interface VideoCacheDB extends DBSchema {
  videos: {
    key: string;  // video_id
    value: {
      metadata: any;
      lastAccessed: number;
      segments: any[];
      sections: any[];
    };
  };
  chatHistory: {
    key: string;  // `${video_id}:${conversation_id}`
    value: {
      messages: any[];
      lastUpdated: number;
    };
  };
  searchResults: {
    key: string;  // `${video_id}:${queryHash}`
    value: {
      results: any;
      timestamp: number;
      query: string;
    };
  };
}

class CacheService {
  private db: IDBPDatabase<VideoCacheDB> | null = null;
  private readonly DB_NAME = 'video-analysis-cache';
  private readonly VERSION = 1;
  private readonly CACHE_DURATION = 7 * 24 * 60 * 60 * 1000; // 7 days

  async initialize(): Promise<void> {
    if (this.db) return;

    this.db = await openDB<VideoCacheDB>(this.DB_NAME, this.VERSION, {
      upgrade(db) {
        // Create video store
        if (!db.objectStoreNames.contains('videos')) {
          const videoStore = db.createObjectStore('videos', { keyPath: 'key' });
          videoStore.createIndex('lastAccessed', 'lastAccessed');
        }

        // Create chat history store
        if (!db.objectStoreNames.contains('chatHistory')) {
          const chatStore = db.createObjectStore('chatHistory', { keyPath: 'key' });
          chatStore.createIndex('lastUpdated', 'lastUpdated');
        }

        // Create search results store
        if (!db.objectStoreNames.contains('searchResults')) {
          const searchStore = db.createObjectStore('searchResults', { keyPath: 'key' });
          searchStore.createIndex('timestamp', 'timestamp');
        }
      },
    });

    // Clean up old entries
    await this.cleanup();
  }

  private async cleanup(): Promise<void> {
    if (!this.db) return;

    const now = Date.now();
    const cutoff = now - this.CACHE_DURATION;

    // Clean up old videos
    const videoTx = this.db.transaction('videos', 'readwrite');
    const videoIndex = videoTx.store.index('lastAccessed');
    let cursor = await videoIndex.openCursor();
    while (cursor) {
      if (cursor.value.lastAccessed < cutoff) {
        await cursor.delete();
      }
      cursor = await cursor.continue();
    }

    // Clean up old chat history
    const chatTx = this.db.transaction('chatHistory', 'readwrite');
    const chatIndex = chatTx.store.index('lastUpdated');
    cursor = await chatIndex.openCursor();
    while (cursor) {
      if (cursor.value.lastUpdated < cutoff) {
        await cursor.delete();
      }
      cursor = await cursor.continue();
    }

    // Clean up old search results
    const searchTx = this.db.transaction('searchResults', 'readwrite');
    const searchIndex = searchTx.store.index('timestamp');
    cursor = await searchIndex.openCursor();
    while (cursor) {
      if (cursor.value.timestamp < cutoff) {
        await cursor.delete();
      }
      cursor = await cursor.continue();
    }
  }

  // Video metadata caching
  async getVideoMetadata(videoId: string): Promise<any | null> {
    if (!this.db) await this.initialize();
    const tx = this.db!.transaction('videos', 'readonly');
    const store = tx.objectStore('videos');
    const result = await store.get(videoId);
    if (result) {
      // Update last accessed time
      await this.updateVideoLastAccessed(videoId);
      return result.metadata;
    }
    return null;
  }

  async setVideoMetadata(videoId: string, metadata: any, segments: any[], sections: any[]): Promise<void> {
    if (!this.db) await this.initialize();
    const tx = this.db!.transaction('videos', 'readwrite');
    const store = tx.objectStore('videos');
    await store.put({
      key: videoId,
      value: {
        metadata,
        segments,
        sections,
        lastAccessed: Date.now(),
      },
    });
  }

  private async updateVideoLastAccessed(videoId: string): Promise<void> {
    if (!this.db) await this.initialize();
    const tx = this.db!.transaction('videos', 'readwrite');
    const store = tx.objectStore('videos');
    const data = await store.get(videoId);
    if (data) {
      data.value.lastAccessed = Date.now();
      await store.put(data);
    }
  }

  // Chat history caching
  async getChatHistory(videoId: string, conversationId: string): Promise<any[] | null> {
    if (!this.db) await this.initialize();
    const tx = this.db!.transaction('chatHistory', 'readonly');
    const store = tx.objectStore('chatHistory');
    const key = `${videoId}:${conversationId}`;
    const result = await store.get(key);
    return result?.messages || null;
  }

  async appendChatMessage(videoId: string, conversationId: string, message: any): Promise<void> {
    if (!this.db) await this.initialize();
    const tx = this.db!.transaction('chatHistory', 'readwrite');
    const store = tx.objectStore('chatHistory');
    const key = `${videoId}:${conversationId}`;
    
    const existing = await store.get(key);
    const messages = existing?.messages || [];
    messages.push(message);

    await store.put({
      key,
      value: {
        messages,
        lastUpdated: Date.now(),
      },
    });
  }

  // Search results caching
  async getSearchResults(videoId: string, query: string): Promise<any | null> {
    if (!this.db) await this.initialize();
    const tx = this.db!.transaction('searchResults', 'readonly');
    const store = tx.objectStore('searchResults');
    const key = `${videoId}:${this.hashQuery(query)}`;
    const result = await store.get(key);
    
    // Check if result is still valid (within 1 hour)
    if (result && Date.now() - result.timestamp < 60 * 60 * 1000) {
      return result.results;
    }
    return null;
  }

  async setSearchResults(videoId: string, query: string, results: any): Promise<void> {
    if (!this.db) await this.initialize();
    const tx = this.db!.transaction('searchResults', 'readwrite');
    const store = tx.objectStore('searchResults');
    const key = `${videoId}:${this.hashQuery(query)}`;
    
    await store.put({
      key,
      value: {
        results,
        query,
        timestamp: Date.now(),
      },
    });
  }

  private hashQuery(query: string): string {
    // Simple hash function for queries
    let hash = 0;
    for (let i = 0; i < query.length; i++) {
      const char = query.charCodeAt(i);
      hash = ((hash << 5) - hash) + char;
      hash = hash & hash;
    }
    return hash.toString(36);
  }

  // Utility methods
  async clearVideoCache(videoId: string): Promise<void> {
    if (!this.db) await this.initialize();
    const tx = this.db!.transaction(['videos', 'chatHistory', 'searchResults'], 'readwrite');
    
    // Clear video metadata
    await tx.objectStore('videos').delete(videoId);
    
    // Clear related chat history
    const chatStore = tx.objectStore('chatHistory');
    let cursor = await chatStore.openCursor();
    while (cursor) {
      if (cursor.key.startsWith(`${videoId}:`)) {
        await cursor.delete();
      }
      cursor = await cursor.continue();
    }
    
    // Clear related search results
    const searchStore = tx.objectStore('searchResults');
    cursor = await searchStore.openCursor();
    while (cursor) {
      if (cursor.key.startsWith(`${videoId}:`)) {
        await cursor.delete();
      }
      cursor = await cursor.continue();
    }
  }

  async clearAll(): Promise<void> {
    if (!this.db) await this.initialize();
    const tx = this.db!.transaction(['videos', 'chatHistory', 'searchResults'], 'readwrite');
    await Promise.all([
      tx.objectStore('videos').clear(),
      tx.objectStore('chatHistory').clear(),
      tx.objectStore('searchResults').clear(),
    ]);
  }
}

// Export singleton instance
export const cacheService = new CacheService(); 

File: backend/app/core/cache.py


from typing import Optional, Dict, Any
import json
from redis.asyncio import Redis
from datetime import timedelta

class VideoCache:
    """Redis-based cache manager for video metadata and processing status"""
    
    def __init__(self, redis_url: str):
        self.redis = Redis.from_url(redis_url, encoding="utf-8", decode_responses=True)
        self.default_ttl = timedelta(days=7)  # Cache videos for 7 days by default
        
    async def get_video_metadata(self, video_id: str) -> Optional[Dict[str, Any]]:
        """Get cached video metadata"""
        data = await self.redis.get(f"video:{video_id}:metadata")
        return json.loads(data) if data else None
    
    async def set_video_metadata(self, video_id: str, metadata: Dict[str, Any], ttl: Optional[timedelta] = None) -> None:
        """Cache video metadata"""
        await self.redis.set(
            f"video:{video_id}:metadata",
            json.dumps(metadata),
            ex=ttl or self.default_ttl
        )
    
    async def get_processing_status(self, video_id: str) -> Optional[str]:
        """Get video processing status"""
        return await self.redis.get(f"video:{video_id}:status")
    
    async def set_processing_status(self, video_id: str, status: str, ttl: Optional[timedelta] = None) -> None:
        """Update video processing status"""
        await self.redis.set(
            f"video:{video_id}:status",
            status,
            ex=ttl or self.default_ttl
        )
    
    async def get_chat_history(self, video_id: str, conversation_id: str) -> list:
        """Get chat history for a conversation"""
        data = await self.redis.get(f"chat:{video_id}:{conversation_id}")
        return json.loads(data) if data else []
    
    async def append_chat_message(self, video_id: str, conversation_id: str, message: Dict[str, Any]) -> None:
        """Append a message to chat history"""
        history = await self.get_chat_history(video_id, conversation_id)
        history.append(message)
        await self.redis.set(
            f"chat:{video_id}:{conversation_id}",
            json.dumps(history),
            ex=self.default_ttl
        )
    
    async def clear_video_cache(self, video_id: str) -> None:
        """Clear all cached data for a video"""
        keys = await self.redis.keys(f"*{video_id}*")
        if keys:
            await self.redis.delete(*keys)
    
    async def get_batch_status(self, batch_id: str) -> Optional[Dict[str, Any]]:
        """Get status of a batch processing job"""
        data = await self.redis.get(f"batch:{batch_id}")
        return json.loads(data) if data else None
    
    async def set_batch_status(self, batch_id: str, status: Dict[str, Any], ttl: Optional[timedelta] = None) -> None:
        """Update batch processing status"""
        await self.redis.set(
            f"batch:{batch_id}",
            json.dumps(status),
            ex=ttl or self.default_ttl
        )
    
    async def close(self) -> None:
        """Close Redis connection"""
        await self.redis.close()

# Create a singleton instance
_cache: Optional[VideoCache] = None

async def get_cache(redis_url: str) -> VideoCache:
    """Get or create cache instance"""
    global _cache
    if _cache is None:
        _cache = VideoCache(redis_url)
    return _cache 